proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,225 @@
+introduced in C++20 to statically validate a type.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. A concept 
+containing the relevant `type_traits` will be applied to each resource declaration,
+and there wil be sufficient context to determine the target IR.
+If DXIL is given, then all of the typed buffer `type_traits` will be applied on each
+typed buffer HLSL resource type. Otherwise, the raw buffer type_traits will be 
+applied to each resource type. If a `type_trait` is not true for the given 
+RET, a corresponding error message will be emitted.
+
+The list of type_traits that will be available for use are described below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_intangible` | An RET should be an arithmetic type, or a bool, or a vector or matrix or UDT containing such types. This is equivalent to validating that the RET is not intangible. |
+| `__builtin_is_homogenous` | A typed buffer RET with the DXIL IR target should never have two different subelement types. |
+
+For the SPIR-V IR target, only `__is_complete_type` and `!__is_intangible` 
+need to be true. When the target IR is DXIL, and the resource is a typed buffer variant,
+`__builtin_is_homogenous` will be used to ensure homogeneity. 
+It will use `BuildFlattenedTypeList` to retrieve a small vector of the subelement types.
+From this subvector, the first element will be compared to all elements in the vector,
+and any mismatches will return false.
+Typed buffer RETs with the DXIL IR target will need have a vector length that is
+at most 4, and the total size in bytes is at most 16. However, type_traits are not
+needed to verify these, since they can be checked directly using template techniques
+and the `sizeof` builtin.
+
+* Examples:
+```
+// targeting DXIL
+struct oneInt {
+	int i;
+};
+struct twoInt {
+   int aa;
+   int ab;
+};
+struct a {
+   oneInt bx;
+   int i;
+};
+struct b;
+struct c {
+  oneInt ca;
+  float1 cb;
+};
+struct d {
+  twoInt x[2];
+  twoInt y[2];
+};
+RWBuffer<double2> r0; // valid - RET fits in 4 32-bit quantities
+RWBuffer<int> r1; // valid
+RWBuffer<float> r2; // valid
+RWBuffer<float4> r3; // valid
+RWBuffer<oneInt> r4; // valid
+RWBuffer<oneInt> r5; // valid - all fields are valid primitive types
+RWBuffer<a> r6; // valid - all leaf types are valid primitive types, and homogenous
+
+// diagnostic: "resource element type 'b' has incomplete definition"
+RWBuffer<b> r7; // invalid - the RET isn't complete, the definition is missing. 
+// the type_trait that would catch this is `__is_complete_type`
+
+// diagnostic: "resource element type 'c' has non-homogenous aggregate type"

```

Comment:
With the proposed concepts-based implementation do you think you'll be able to get these exact diagnostics?

Reply:
No, I don't think the diagnostics are any longer controllable.
For example, if an RET is a vector of 8 doubles, we'd expect these diagnostics:

```
<source>:41:5: error: constraints not satisfied for class template 'RWBuffer' [with T = double __attribute__((ext_vector_type(8)))]
   41 |     RWBuffer<vector<double, 8>> Buf;
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:29:32: note: because '!is_vector_type<double __attribute__((ext_vector_type(8)))>::value' evaluated to false
   29 | template<typename T> requires (!is_vector_type<T>::value) || (is_spirv_target || is_valid_line_vector<T>) && __is_complete_type(T)
      |                                ^
<source>:29:63: note: and 'is_spirv_target' evaluated to false
   29 | template<typename T> requires (!is_vector_type<T>::value) || (is_spirv_target || is_valid_line_vector<T>) && __is_complete_type(T)
      |                                                               ^
<source>:29:82: note: and 'double __attribute__((ext_vector_type(8)))' (vector of 8 'double' values) does not satisfy 'is_valid_line_vector'
   29 | template<typename T> requires (!is_vector_type<T>::value) || (is_spirv_target || is_valid_line_vector<T>) && __is_complete_type(T)
      |                                                                                  ^
<source>:27:60: note: because 'sizeof(double __attribute__((ext_vector_type(8)))) <= 16' (64 <= 16) evaluated to false
   27 | concept is_valid_line_vector = is_vector_type<T>::value && sizeof(T) <= 16 && vector_type_info<T>::Size <= 4;
      | 
```

I'll remove the expected diagnostics.

Reply:
We should have some text in here saying that we're going to rely on the standard diagnostics emitted when constraints aren't satisfied.